lib/chains/videoChain.js

```diff
@@ -1,8 +1,234 @@
-const {element} = require('./elementChain');
+const {isNil} = require('ramda');
 
-const videoChain = () => element({video: true});
+// Import utils
+const makeChain = require('../utils/makeChain');
+
+// Import chain composers
+const {
+	notComposer,
+	existComposer,
+	matchComposer,
+	matchRepoComposer,
+	matchJSComposer,
+	// matchBrightScriptComposer,
+	timeoutComposer,
+	makeToStringComposer,
+	makeThenComposer,
+	abandonComposer,
+	cloneComposer,
+	assertComposer,
+	gettersComposer,
+	makeToJSONComposer,
+	isPlayingComposer,
+	isStoppedComposer,
+	isPausedComposer,
+	visibleComposer,
+} = require('../composers');
+const {
+	element,
+	elementExist,
+	elementNotExist,
+	elementVisible,
+	elementMatchProps,
+	elementMatchJS,
+	elementNotMatchJS,
+	elementMatchBRS,
+	elementNotMatchBRS,
+} = require('../texts');
+const {SUBJ_COMPARATOR} = require('../constants/comparator');
+const {
+	processServerResponse,
+	getRequestType,
+	applyTimeout,
+	applyNegation,
+} = require('../utils/socketChainHelper');
+const {
+	SUBJ_COMPARATOR: SUBJ_COMPARATOR_M,
+	VALUE: VALUE_M,
+} = require('../mappings');
+const {VALUE} = require('../constants/element');
+const logger = require('../utils/logger');
+const getOpType = require('../utils/opType').getOpType;
+const {placeholdEmpty} = require('../utils/stringUtils');
+
+/**
+ * Returns human readable description of the video chain.
+ * @param data - chain data
+ * @param nameOnly - if true only element name will be returned (like: {"apiId" : "boo"})
+ * @returns {*}
+ */
+const toString = (data, nameOnly = false) => {
+	const elName = 'video';
+
+	if (nameOnly)
+		return elName;
+
+	let out = element(elName);
+
+	if (data.comparator) {
+		switch (data.comparator.type) {
+			case SUBJ_COMPARATOR.VISIBLE:
+				out = elementVisible(elName);
+				break;
+			case SUBJ_COMPARATOR.EXIST:
+				out = (data.isNegated ? elementNotExist : elementExist)(elName);
+				break;
+			case SUBJ_COMPARATOR.MATCH_JS:
+				out = (data.isNegated ? elementNotMatchJS : elementMatchJS)(elName, data.comparator.val);
+				break;
+			case SUBJ_COMPARATOR.MATCH_BRS:
+				out = (data.isNegated ? elementNotMatchBRS : elementMatchBRS)(elName, data.comparator.val);
+				break;
+			default: {
+				const matches = data.comparator.props.map(
+					one => `  ${one.name} ${one.type} ${placeholdEmpty(VALUE_M[one.val] || one.val)}`
+				);
+
+				out = elementMatchProps(elName, '\n' + matches.join(',\n'));
+			}
+		}
+	}
+
+	return out;
+};
+const beforeSendMsg = (data) => logger.log(getOpType(data), toString(data));
+
+const toJSON = data => {
+	const type = getRequestType(data);
+	const socketMessage = {type};
+	const subject = {
+		type: 'video',
+	};
+
+	// query
+	if (type === 'query') {
+		socketMessage.subject = {
+			type: 'elementProps',
+			selector: {video: true},
+		};
+	}
+
+	// video subject
+	if (data.comparator) {
```

Comment:
this block of code (and at 69 line) are duplicates from elementChain.js. Is it possible to reuse them?

Reply:
@AntonSBilan how about moving it to some helper/decorator?

Reply:
Yes, there is some duplicated lines of code, but I thought the idea of this separation is to make doing changes in element and video more clean and easier. I'm afraid that incapsulating this logic may return us back to what we had

Reply:
Ok, I will update it

Reply:
The reason we are separating this two chain subject is because they have different API. I.e. element should not have chain methods like `isPlaying`, while video should not have `setText`. So, they are different enough. However, this does not forbid us from having some code moved to helpers and reused in both cases to keep it DRY. It really depend on how much code are we talking about. If it's about 10-20 lines - probably not worth it. Otherwise I would create a helper.